identity-core/src/crypto/merkle_key/verifier.rs

```diff
@@ -74,63 +59,46 @@ where
   }
 }
 
-impl<S, D> Verifier<'_, S, D>
+// =============================================================================
+// =============================================================================
+
+/// A signature verification helper for Merkle Key Collection Signatures.
+#[derive(Clone)]
+pub struct MerkleVerifier<D, S>
 where
-  S: MerkleSignature,
   D: MerkleDigest,
+  S: MerkleSignature,
 {
-  fn decompose_public_key(&self, digest: &D) -> Result<Hash<D>> {
-    let (tag_s, tag_d): (MerkleTag, MerkleTag) = MerkleKey::extract_tags(&self.merkle_key)?;
-
-    // Validate the signature algorithm tag
-    if tag_s != self.suite.tag() {
-      return Err(Error::InvalidMerkleKeyTag(Some(tag_d)));
-    }
-
-    // Validate the digest algorithm tag
-    if tag_d != digest.tag() {
-      return Err(Error::InvalidMerkleKeyTag(Some(tag_d)));
-    }
-
-    // Extract and return the Merkle root hash
-    self
-      .merkle_key
-      .get(2..)
-      .and_then(Hash::from_slice)
-      .ok_or(Error::InvalidKeyFormat)
-  }
+  marker_d: PhantomData<D>,
+  marker_s: PhantomData<S>,
 }
 
-impl<S, D> SignatureName for Verifier<'_, S, D>
+impl<D, S> Named for MerkleVerifier<D, S>
 where
   D: MerkleDigest,
+  S: MerkleSignature,
 {
-  fn name(&self) -> String {
-    MerkleKey::TYPE_SIG.to_string()
-  }
+  const NAME: &'static str = MerkleKey::TYPE_SIG;
 }
 
-impl<S, D> SignatureVerify for Verifier<'_, S, D>
+impl<'key, D, S> Verifier<VerificationKey<'key>> for MerkleVerifier<D, S>
 where
-  S: MerkleSignature + SignatureVerify,
   D: MerkleDigest,
+  S: MerkleSignature,
+  S: Verify,
+  S: Verify<Public = [u8]>,
 {
-  fn verify(&self, data: &dyn Serialize, signature: &SignatureValue, public: &[u8]) -> Result<()> {
-    // Merkle Key Collection signatures store their public key values
-    // alongside their proofs (in the signature value). The `public` value
-    // **SHOULD NOT** be provided as it will be extracted and verified
-    // according to the decoded proof and revocation set.
-    if !public.is_empty() {
-      return Err(Error::InvalidKeyFormat);
-    }
-
+  fn verify<X>(data: &X, signature: &SignatureValue, public: &VerificationKey<'key>) -> Result<()>
+  where
+    X: Serialize,
+  {
     let mut digest: D = D::new();
 
     let (target, proof, signature): _ = expand_signature_value(signature)?;
```

Comment:
Do we need this underscore here?

Reply:
probably not. going to add this to a cleanup task since `_` is used unnecessarily in dozens of places